src/lazy.jl

```diff
@@ -0,0 +1,722 @@
+        end
+    end)
+end
+
+# core JSON array parsing function
+# takes a `keyvalfunc` that is applied to each index => value element
+# `keyvalfunc` is provided a Int => LazyValue pair
+# applyeach always requires a key-value pair function
+# so we use the index as the key
+# returns a `pos` value that notes the next position where parsing should continue
+# this is essentially the `StructUtils.applyeach` implementation for LazyValues w/ type ARRAY
+function applyarray(keyvalfunc, x::LazyValues)
+    pos = getpos(x)
+    buf = getbuf(x)
+    len = getlength(buf)
+    opts = getopts(x)
+    jsonlines = opts.jsonlines
+    b = getbyte(buf, pos)
+    if !jsonlines
+        if b != UInt8('[')
+            error = ExpectedOpeningArrayChar
+            @goto invalid
+        end
+        pos += 1
+        @nextbyte
+        b == UInt8(']') && return pos + 1
+    else
+        # for jsonlines, we need to make sure that recursive
+        # lazy values *don't* consider individual lines *also*
+        # to be jsonlines
+        opts = LazyOptions(; allownan=opts.allownan, ninf=opts.ninf, inf=opts.inf, nan=opts.nan, jsonlines=false)
+    end
+    i = 1
+    while true
+        # we're now positioned at the start of the value
+        val = lazy(buf, pos, len, b, opts)
+        ret = keyvalfunc(i, val)
+        ret isa StructUtils.EarlyReturn && return ret
+        # if keyvalfunc didn't materialize `val` and return an
+        # updated `pos`, then we need to skip val ourselves
+        # WARNING: parsing can get corrupted if random Int values are returned from keyvalfunc
+        pos = (ret isa Int && ret > pos) ? ret : skip(val)
+        if jsonlines
+            @jsonlines_checks
+        else
+            @nextbyte
+            if b == UInt8(']')
+                return pos + 1
+            elseif b != UInt8(',')
+                error = ExpectedComma
+                @goto invalid
+            end
+            pos += 1 # move past ','
+            @nextbyte
+        end
+        i += 1
+    end
+
+@label invalid
+    invalid(error, buf, pos, "array")
+end
+
+# temporary string type to enable deferrment of string allocation in certain cases (like navigating a lazy structure)
+struct PtrString
+    ptr::Ptr{UInt8}
+    len::Int
+    escaped::Bool
+end
+
+if VERSION < v"1.11"
+    mem(n) = Vector{UInt8}(undef, n)
+    _tostr(m::Vector{UInt8}, slen) = ccall(:jl_array_to_string, Ref{String}, (Any,), resize!(m, slen))
+else
+    mem(n) = Memory{UInt8}(undef, n)
+    _tostr(m::Memory{UInt8}, slen) = ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), m, slen)
+end
+
+function Base.convert(::Type{String}, x::PtrString)
+    if x.escaped
+        m = mem(x.len)
+        slen = GC.@preserve m unsafe_unescape_to_buffer(x.ptr, x.len, pointer(m))
+        return _tostr(m, slen)
+    end
+    return unsafe_string(x.ptr, x.len)
+end
+
+Base.convert(::Type{Symbol}, x::PtrString) = ccall(:jl_symbol_n, Ref{Symbol}, (Ptr{UInt8}, Int), x.ptr, x.len)
```

Comment:
`jl_symbol_n` is not public.

Reply:
I've opened https://github.com/JuliaLang/julia/issues/58520 to discuss exposing it via `unsafe_symbol`.